File: src/data/courses/dsa/dynamic-programming/classic-dp/burst-balloons/code/python/burst_balloons.py
class Solution:
    def maxCoins(self, nums) -> int:
        a = [1] + nums + [1]; n = len(nums)
        dp = [[0]*(n+2) for _ in range(n+2)]
        for length in range(1, n+1):
            for i in range(1, n-length+2):
                j = i+length-1
                for k in range(i, j+1):
                    dp[i][j] = max(dp[i][j], dp[i][k-1]+a[i-1]*a[k]*a[j+1]+dp[k+1][j])
        return dp[1][n]

File: src/data/courses/dsa/dynamic-programming/classic-dp/burst-balloons/code/java/BurstBalloons.java
class Solution {
    public int maxCoins(int[] nums) {
        int n=nums.length; int[] a=new int[n+2]; a[0]=a[n+1]=1;
        for(int i=0;i<n;i++) a[i+1]=nums[i];
        int[][] dp=new int[n+2][n+2];
        for(int len=1;len<=n;len++) for(int i=1;i<=n-len+1;i++){int j=i+len-1;
            for(int k=i;k<=j;k++) dp[i][j]=Math.max(dp[i][j],dp[i][k-1]+a[i-1]*a[k]*a[j+1]+dp[k+1][j]);}
        return dp[1][n];
    }
}